seal_csv_to_gpkg.py

```diff
@@ -0,0 +1,58 @@
+from dataclasses import dataclass
+
+import fsspec
+from apache_beam.options.pipeline_options import PipelineOptions
+import apache_beam as beam
+
+
+INPUT_URL = "https://arcticdata.io/metacat/d1/mn/v2/object/urn%3Auuid%3A31162eb9-7e3b-4b88-948f-f4c99f13a83f"
+OUTPUT_PATH = "foo.gpkg"
+
+# This object lets us set various options for our pipeline, such as the pipeline
+# runner that will execute our pipeline and any runner-specific configuration
+# required by the chosen runner.
+beam_options = PipelineOptions(
+    runner='DirectRunner',
+    project='seal_tags',
+    # TODO: do we have to specify this? Could be generated automatically from
+    # `project` + unique key?
+    # job_name='seal_tags_job',
+    temp_location='./seal_tags-tmp',
+)
+
+def write_csv(pcoll):
+    import pandas as pd
+    import geopandas
+
+    local_fs = pcoll # <OpenFile 'https://arcticdata.io/metacat/d1/mn/v2/object/urn%3Auuid%3A31162eb9-7e3b-4b88-948f-f4c99f13a83f'>
+
+    with local_fs as f:
+        df = pd.read_csv(f)
+        geom = geopandas.points_from_xy(df.Longitude, df.Latitude)
+        geo_df = geopandas.GeoDataFrame(df, geometry=geom)
+        geo_df.to_file(OUTPUT_PATH, crs="EPSG:4326")
+
+    return OUTPUT_PATH
+
+@dataclass
+class WriteCSV(beam.PTransform):
+    def expand(self, pcoll):
+        return pcoll | beam.Map(write_csv)
+
+
+def open_with_fsspec(fname):
+    return fsspec.open(fname, mode="rb")
+
+@dataclass
+class OpenURLWithFSSpec(beam.PTransform):
+    def expand(self, pcoll):
+        return pcoll | beam.Map(open_with_fsspec)
+
+recipe = (
+    beam.Create([INPUT_URL])
```

Comment:
We've discussed this already, just re-posting for visibility.

Is the more canonical beam way to do this to pass an fsspec object to a beam.io transformer?

Reply:
Relevant section in the programming guide: https://beam.apache.org/documentation/programming-guide/#creating-a-pcollection

> You create a PCollection by either reading data from an external source using Beam’s [Source API](https://beam.apache.org/documentation/programming-guide/#pipeline-io), or you can create a PCollection of data stored in an in-memory collection class in your driver program. The former is typically how a production pipeline would ingest data; Beam’s Source APIs contain adapters to help you read from external sources like large cloud-based files, databases, or subscription services. The latter is primarily useful for testing and debugging purposes.

...

> To create a PCollection from an in-memory list, you use the Beam-provided Create transform. Apply this transform directly to your Pipeline object itself.

So, I think we're doing the 'testing/debugging' approach. We create an in-memory PCollection of len 1 (the input URL), which then gets read w/ fsspec using the next transform in the chain.

I imagine in the future we might have some kind of reader for DataONE metadata that we'd pass args to (e.g., dataset ID) and then get a PCollection of data files (maybe fsspec objects) for that dataset which would be passed to subsequent steps in the chain to act on.

Reply:
What I'm getting at is I don't think we need ReadWithFsspec because I think we can do it with built-in `beam.io` PTransforms. The `beam.Create` method doesn't seem to be gaining us anything here.